knimin/static/js/platemapper.js

```diff
@@ -0,0 +1,506 @@
+/**
+ *
+ * New widget to allow show the study labels in the dropdown menu
+ *
+ * Based on the example in https://jqueryui.com/autocomplete/#categories
+ *
+ */
+$( function() {
+  $.widget( "custom.catcomplete", $.ui.autocomplete, {
+    _create: function() {
+      this._super();
+      this.widget().menu( "option", "items", "> :not(.ui-autocomplete-category)" );
+    },
+    _renderMenu: function( ul, items ) {
+      var that = this,
+          currentCategory = "";
+      $.each( items, function( index, item ) {
+        var li;
+        if ( item.category != currentCategory ) {
+          $('<li>').addClass('pm_study_autocomplete').css({'background-color': item.color}).append(item.category).appendTo(ul);
+          currentCategory = item.category;
+        }
+        li = that._renderItemData( ul, item );
+        if ( item.category ) {
+          li.attr( "aria-label", item.category + " : " + item.label );
+        }
+        });
+      }
+    });
+});
+
+/**
+ *
+ * @class PlateMap
+ *
+ * Represents a Plate
+ *
+ * @param {string} target The name of the target container for the plate map
+ * @param {int} rows The number of rows in the plate
+ * @param {int} cols The number of columns in the plate
+ * @param {string[]} samples The list of samples to populate the autocomplete
+ *
+ * @return {PlateMap}
+ * @constructs PlateMap
+ *
+ **/
+function PlateMap(target, plate_id) {
+  var obj = this;
+
+  // Store the plate ID and the target DOM element
+  this.plateId = plate_id
+  this.target = $('#' + target)
+
+  // Retrieve the plate information from the server
+  $.get('/pm_sample_plate?plate_id=' + plate_id, function (data) {
+    obj.initialize(data);
+  })
+    .fail(function (jqXHR, textStatus, errorThrown) {
+      if(jqXHR.status === 404) {
+        var data = $.parseJSON(jqXHR.responseText)
+        $('<h3 class="warning">').html(data.message).appendTo(obj.target);
+      } else {
+        $('<div>').html(jqXHR.responseText).appendTo(obj.target);
```

Comment:
As a presentation suggestion, if this is an error, or something non-immediately interpretable (developer, communication error), I would hide this error and just show a sentence that reads "An unexpected error occurred >" then if the user clicks on the `>` the full error message or text is shown.